test/cpp/tensorexpr/test_loopnest.cpp

```diff
@@ -546,5 +547,224 @@ void testScheduleDynamicShape2D() {
   testWithSize(37, 11);
 }
 
+static std::unordered_map<const Buf*, TensorAccessBoundsInfo>
+convertBoundsInfoToMap(const std::vector<TensorAccessBoundsInfo>& v) {
+  std::unordered_map<const Buf*, TensorAccessBoundsInfo> res;
+  for (const auto& el : v) {
+    res[el.buf] = el;
+  }
+  return res;
+}
+
+static void verifyConstBounds(
+    const TensorAccessBoundsInfo& access_info,
+    const std::vector<std::pair<int, int>>& ref) {
+  size_t ndim = ref.size();
+  ASSERT_EQ(access_info.start.size(), ndim);
+  ASSERT_EQ(access_info.stop.size(), ndim);
+  for (size_t i = 0; i < ndim; i++) {
+    if (ref[i].first >= 0) { // Negative values are used to skip the check
+      auto start_imm = dynamic_cast<const IntImm*>(access_info.start[i]);
+      ASSERT_TRUE(start_imm);
+      ASSERT_EQ(start_imm->value(), ref[i].first);
+    }
+    if (ref[i].second >= 0) {
+      auto stop_imm = dynamic_cast<const IntImm*>(access_info.stop[i]);
+      ASSERT_TRUE(stop_imm);
+      ASSERT_EQ(stop_imm->value(), ref[i].second);
+    }
+  }
+}
+
+void testBoundsInference_1() {
+  // Verify that bounds inference works for the following example:
+  // for i in 0..100:
+  //   b[i] = a[i]
+  // For this loop bounds inference should yield the following:
+  // {{b, kStore, 0, 99}, {a, kLoad, 0, 99}}
+  KernelScope kernel_scope;
+  ExprHandle n(100);
+  Buffer a(BufHandle("a", {n}), kFloat);
+  Tensor* b =
+      Compute("b", {{n, "i"}}, [&](const VarHandle& i) { return a(i); });
+  LoopNest l({b});
+  const std::vector<TensorAccessBoundsInfo>& bounds_info =
```

Comment:
As for the interface of bounds inference module, those API should be abstracted to encapsulate its implementations.

TensorAccessBounds can be an object itself. Whether the data structure is a vector, or a map can be hidden away from its user.

"bounds_info_map.at(a.data())" Can be replaced by a convenience function "tensor_bounds.find(a)".

Reply:
The API is to return a collection (in this implementation a vector) of `TensorAccessBoundsInfo` objects - I don't see a value in wrapping the collection into a separate class when it's nothing else than just a vector. Note that everyone who reads this code would know what `std::vector` is, but noone except me would know what `TensorAccessBounds` class is.

Reply:
Hiding away the implementation details is a key idea of software abstraction. The interface of this module should reflect its logic usage. If we change the internal data structure, we don't have to change all the call sites, as this might be used in many places. Also we don't have to advertise to every developer the change of the implementation data structure. So interface between the modules can stay relatively stable.